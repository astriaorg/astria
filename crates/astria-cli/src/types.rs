use color_eyre::eyre;
use serde::{
    Deserialize,
    Serialize,
};

use crate::cli::rollup::{
    ConfigCreateArgs,
    GenesisAccountArg,
};

/// Rollup contains the deployment config for a rollup
#[derive(Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Rollup {
    // TODO - namespace
    #[serde(rename = "config")]
    pub(crate) deployment_config: RollupDeploymentConfig,
}

impl TryFrom<&ConfigCreateArgs> for Rollup {
    type Error = eyre::Report;

    fn try_from(args: &ConfigCreateArgs) -> eyre::Result<Self> {
        let deployment_config = RollupDeploymentConfig::try_from(args)?;

        Ok(Self {
            deployment_config,
        })
    }
}

impl TryInto<String> for Rollup {
    type Error = eyre::Report;

    /// Serializes Rollup to a yaml string
    fn try_into(self) -> eyre::Result<String> {
        let yaml_str = serde_yaml::to_string(&self)?;
        Ok(yaml_str)
    }
}

/// Describes a rollup deployment config. Serializes to a yaml file for usage with Helm,
/// thus the `rename_all = "camelCase"` naming convention.
#[derive(Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RollupDeploymentConfig {
    #[serde(rename = "useTTY")]
    use_tty: bool,
    log_level: String,
    rollup: RollupConfig,
    sequencer: SequencerConfig,
}

impl RollupDeploymentConfig {
    #[must_use]
    pub fn get_filename(&self) -> String {
        format!("{}-rollup-conf.yaml", self.rollup.name)
    }

    #[must_use]
    pub fn get_chart_release_name(&self) -> String {
        format!("{}-rollup", self.rollup.name)
    }

    #[must_use]
    pub fn get_rollup_name(&self) -> String {
        self.rollup.name.clone()
    }
}

impl TryFrom<&ConfigCreateArgs> for RollupDeploymentConfig {
    type Error = eyre::Report;

    fn try_from(args: &ConfigCreateArgs) -> eyre::Result<Self> {
        let chain_id = args
            .chain_id
            .clone()
            .unwrap_or(format!("{}-chain", args.name));

        let sequencer_initial_block_height = args.sequencer_initial_block_height.unwrap_or({
            // TODO - get current block height from sequencer
            0
        });

        let genesis_accounts = args
            .genesis_accounts
            .clone()
            .into_iter()
            .map(GenesisAccount::from)
            .collect();

        Ok(Self {
            use_tty: args.use_tty,
            log_level: args.log_level.clone(),
            rollup: RollupConfig {
                name: args.name.clone(),
                chain_id,
                network_id: args.network_id,
                skip_empty_blocks: args.skip_empty_blocks,
                genesis_accounts,
            },
            sequencer: SequencerConfig {
                initial_block_height: sequencer_initial_block_height,
                websocket: args.sequencer_websocket.clone(),
                rpc: args.sequencer_rpc.clone(),
            },
        })
    }
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RollupConfig {
    name: String,
    chain_id: String,
    network_id: u64,
    skip_empty_blocks: bool,
    genesis_accounts: Vec<GenesisAccount>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct GenesisAccount {
    address: String,
    // NOTE - string because yaml will serialize large ints w/ scientific notation
    balance: String,
}

impl From<GenesisAccountArg> for GenesisAccount {
    fn from(arg: GenesisAccountArg) -> Self {
        Self {
            address: arg.address,
            balance: arg.balance.to_string(),
        }
    }
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct SequencerConfig {
    initial_block_height: u64,
    websocket: String,
    rpc: String,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::cli::rollup::{
        ConfigCreateArgs,
        GenesisAccountArg,
    };

    #[test]
    fn test_from_cli_args() -> eyre::Result<()> {
        // Case 1: All args provided
        let args1 = ConfigCreateArgs {
            use_tty: true,
            log_level: "debug".to_string(),
            name: "rollup1".to_string(),
            chain_id: Some("chain1".to_string()),
            network_id: 1,
            skip_empty_blocks: true,
            genesis_accounts: vec![
                GenesisAccountArg {
                    address: "0xA5TR14".to_string(),
                    balance: 1_000_000_000_000_000_000,
                },
                GenesisAccountArg {
                    address: "0x420XYZ69".to_string(),
                    balance: 420,
                },
            ],
            sequencer_initial_block_height: Some(10),
            sequencer_websocket: "ws://localhost:8080".to_string(),
            sequencer_rpc: "http://localhost:8081".to_string(),
        };

        let expected_config1 = Rollup {
            deployment_config: RollupDeploymentConfig {
                use_tty: true,
                log_level: "debug".to_string(),
                rollup: RollupConfig {
                    name: "rollup1".to_string(),
                    chain_id: "chain1".to_string(),
                    network_id: 1,
                    skip_empty_blocks: true,
                    genesis_accounts: vec![
                        GenesisAccount {
                            address: "0xA5TR14".to_string(),
                            balance: "1000000000000000000".to_string(),
                        },
                        GenesisAccount {
                            address: "0x420XYZ69".to_string(),
                            balance: "420".to_string(),
                        },
                    ],
                },
                sequencer: SequencerConfig {
                    initial_block_height: 10,
                    websocket: "ws://localhost:8080".to_string(),
                    rpc: "http://localhost:8081".to_string(),
                },
            },
        };

        let result1 = Rollup::try_from(&args1)?;
        assert_eq!(result1, expected_config1);

        // Case 2: No `Option` wrapped args provided. Tests defaults that are decided
        //  explicitly in the `try_from` impl.
        // NOTE - there are some defaults that are handled in the arg struct,
        //  like the sequencer ws and rpc urls, so we still must pass them in here.
        let args2 = ConfigCreateArgs {
            use_tty: false,
            log_level: "info".to_string(),
            name: "rollup2".to_string(),
            chain_id: None,
            network_id: 2,
            skip_empty_blocks: false,
            genesis_accounts: vec![GenesisAccountArg {
                address: "0xA5TR14".to_string(),
                balance: 10000,
            }],
            sequencer_initial_block_height: None,
            sequencer_websocket: "ws://localhost:8082".to_string(),
            sequencer_rpc: "http://localhost:8083".to_string(),
        };

        let expected_config2 = Rollup {
            deployment_config: RollupDeploymentConfig {
                use_tty: false,
                log_level: "info".to_string(),
                rollup: RollupConfig {
                    name: "rollup2".to_string(),
                    chain_id: "rollup2-chain".to_string(), // Derived from name
                    network_id: 2,
                    skip_empty_blocks: false,
                    genesis_accounts: vec![GenesisAccount {
                        address: "0xA5TR14".to_string(),
                        balance: "10000".to_string(),
                    }],
                },
                sequencer: SequencerConfig {
                    initial_block_height: 0, // Default value
                    websocket: "ws://localhost:8082".to_string(),
                    rpc: "http://localhost:8083".to_string(),
                },
            },
        };

        let result2 = Rollup::try_from(&args2)?;
        assert_eq!(result2, expected_config2);

        Ok(())
    }
}
