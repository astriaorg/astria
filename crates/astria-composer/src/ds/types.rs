use ethers::types::Transaction as EthersTx;

#[non_exhaustive]
pub enum RollupTx {
    EthersTx(EthersTx),
}

pub(crate) type RollupChainId = String;
pub type RollupTxExt = (RollupTx, RollupChainId);
