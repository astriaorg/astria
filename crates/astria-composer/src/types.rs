use serde::{
    Deserialize,
    Serialize,
};

#[derive(Debug, Deserialize, Serialize, PartialEq)]
pub struct ChainId(String);
