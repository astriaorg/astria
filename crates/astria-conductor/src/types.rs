use astria_sequencer_types::{
    ChainId,
    RawSequencerBlockData,
    SequencerBlockData,
};
use tendermint::{
    block::Header,
    Hash,
};

/// `SequencerBlockSubset` is a subset of a SequencerBlock that contains
/// information required for transaction data verification, and the transactions
/// for one specific rollup.
#[derive(Clone, Debug)]
pub(crate) struct SequencerBlockSubset {
    pub(crate) block_hash: Hash,
    pub(crate) header: Header,
    pub(crate) rollup_transactions: Vec<Vec<u8>>,
}

impl SequencerBlockSubset {
    pub(crate) fn from_sequencer_block_data(
        data: SequencerBlockData,
        chain_id: &ChainId,
    ) -> Option<Self> {
        // we don't need to verify the action tree root here,
        // as [`SequencerBlockData`] would not be constructable
        // if it was invalid
        let RawSequencerBlockData {
            block_hash,
            header,
            last_commit: _,
            mut rollup_data,
            ..
        } = data.into_raw();

        let rollup_transactions = rollup_data.remove(chain_id)?;
        Some(Self {
            block_hash,
            header,
            rollup_transactions,
        })
    }
}
