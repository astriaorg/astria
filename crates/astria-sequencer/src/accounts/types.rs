use borsh::{
    BorshDeserialize,
    BorshSerialize,
};
use serde::{
    Deserialize,
    Serialize,
};

/// Address represents an account address.
#[derive(Clone, BorshSerialize, BorshDeserialize, Serialize, Deserialize, PartialEq, Eq, Debug)]
pub(crate) struct Address(String);

impl AsRef<str> for Address {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

impl From<&str> for Address {
    fn from(s: &str) -> Self {
        Self(s.to_string())
    }
}

impl Address {
    pub(crate) fn to_str(&self) -> &str {
        &self.0
    }
}

/// Balance represents an account balance.
#[derive(
    Clone,
    Copy,
    BorshSerialize,
    BorshDeserialize,
    Serialize,
    Deserialize,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Debug,
)]
pub(crate) struct Balance(u128);

impl Balance {
    pub(crate) fn into_inner(self) -> u128 {
        self.0
    }
}

impl From<u128> for Balance {
    fn from(n: u128) -> Self {
        Self(n)
    }
}

impl std::ops::Add<u128> for Balance {
    type Output = Self;

    fn add(self, rhs: u128) -> Self::Output {
        Self(self.0 + rhs)
    }
}

impl std::ops::Add for Balance {
    type Output = Self;

    fn add(self, rhs: Self) -> Self::Output {
        Self(self.0 + rhs.0)
    }
}

impl std::ops::Sub for Balance {
    type Output = Self;

    fn sub(self, rhs: Self) -> Self::Output {
        Self(self.0 - rhs.0)
    }
}

impl std::ops::Sub<u128> for Balance {
    type Output = Self;

    fn sub(self, rhs: u128) -> Self::Output {
        Self(self.0 - rhs)
    }
}

// Nonce represents an account nonce.
#[derive(
    Clone,
    Copy,
    BorshSerialize,
    BorshDeserialize,
    Serialize,
    Deserialize,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Debug,
)]
pub(crate) struct Nonce(u32);

impl Nonce {
    pub(crate) fn into_inner(self) -> u32 {
        self.0
    }
}

impl From<u32> for Nonce {
    fn from(n: u32) -> Self {
        Self(n)
    }
}

impl std::ops::Add for Nonce {
    type Output = Self;

    fn add(self, rhs: Self) -> Self::Output {
        Self(self.0 + rhs.0)
    }
}

impl std::cmp::PartialEq<u32> for Nonce {
    fn eq(&self, other: &u32) -> bool {
        self.0 == *other
    }
}
