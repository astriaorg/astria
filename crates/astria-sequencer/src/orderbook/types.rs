use borsh::{BorshDeserialize, BorshSerialize};
use serde::{Deserialize, Serialize};
// Fix import paths
use astria_core::generated::astria::primitive::v1::Address;
use astria_core::generated::astria::primitive::v1::Uint128;

// Order side (buy or sell)
#[derive(Debug, Clone, Copy, PartialEq, Eq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
pub enum OrderSide {
    Buy,
    Sell,
}

// Order type (limit, market, etc.)
#[derive(Debug, Clone, Copy, PartialEq, Eq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
pub enum OrderType {
    Limit,
    Market,
}

// Order time in force
#[derive(Debug, Clone, Copy, PartialEq, Eq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
pub enum OrderTimeInForce {
    GoodTillCancelled, // Order remains active until explicitly cancelled
    ImmediateOrCancel, // Execute as much as possible immediately, cancel remaining
    FillOrKill,        // Fill entire order or cancel entirely
}

// Type alias for component module
pub type TimeInForce = OrderTimeInForce;

// Market information
#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
pub struct Market {
    pub id: String,
    pub base_asset: String,
    pub quote_asset: String,
    pub tick_size: String,
    pub lot_size: String,
    pub paused: bool,
}

// Order structure
#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
pub struct Order {
    pub id: String,
    pub owner: String,
    pub market: String,
    pub side: OrderSide,
    pub order_type: OrderType,
    pub price: String,
    pub quantity: String,
    pub remaining_quantity: String,
    pub created_at: u64,
    pub time_in_force: OrderTimeInForce,
}

// Order book entry showing aggregate quantity at a price level
#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
pub struct OrderbookEntry {
    pub price: String,
    pub quantity: String,
    pub order_count: u32,
}

// Complete order book for a market
#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
pub struct Orderbook {
    pub market: String,
    pub bids: Vec<OrderbookEntry>,
    pub asks: Vec<OrderbookEntry>,
}

// Record of a completed trade/match
#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
pub struct Trade {
    pub id: String,
    pub market: String,
    pub price: String,
    pub quantity: String,
    pub maker_order_id: String,
    pub taker_order_id: String,
    pub maker_side: OrderSide,
    pub timestamp: u64,
}

// Compatibility layer for OrderMatch
#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
pub struct OrderMatch {
    pub id: String,
    pub market: String,
    pub price: String,
    pub quantity: String,
    pub maker_order_id: String,
    pub taker_order_id: String,
    pub taker_side: OrderSide,
    pub timestamp: u64,
}

impl From<OrderMatch> for Trade {
    fn from(match_info: OrderMatch) -> Self {
        Self {
            id: match_info.id,
            market: match_info.market,
            price: match_info.price,
            quantity: match_info.quantity,
            maker_order_id: match_info.maker_order_id,
            taker_order_id: match_info.taker_order_id,
            maker_side: match opposing_side(match_info.taker_side) {
                OrderSide::Buy => OrderSide::Buy,
                OrderSide::Sell => OrderSide::Sell,
            },
            timestamp: match_info.timestamp,
        }
    }
}

impl From<Trade> for OrderMatch {
    fn from(trade: Trade) -> Self {
        Self {
            id: trade.id,
            market: trade.market,
            price: trade.price,
            quantity: trade.quantity,
            maker_order_id: trade.maker_order_id,
            taker_order_id: trade.taker_order_id,
            taker_side: match opposing_side(trade.maker_side) {
                OrderSide::Buy => OrderSide::Buy,
                OrderSide::Sell => OrderSide::Sell,
            },
            timestamp: trade.timestamp,
        }
    }
}

// Get the opposing side of an order
pub fn opposing_side(side: OrderSide) -> OrderSide {
    match side {
        OrderSide::Buy => OrderSide::Sell,
        OrderSide::Sell => OrderSide::Buy,
    }
}

// Create order action
#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
pub struct CreateOrderAction {
    pub market: String,
    pub side: OrderSide,
    pub order_type: OrderType,
    pub price: String,
    pub quantity: String,
    pub time_in_force: OrderTimeInForce,
    pub fee_asset: String,
}

// Cancel order action
#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
pub struct CancelOrderAction {
    pub order_id: String,
    pub fee_asset: String,
}

// Create market action
#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
pub struct CreateMarketAction {
    pub market_id: String,
    pub base_asset: String,
    pub quote_asset: String,
    pub tick_size: String,
    pub lot_size: String,
    pub fee_asset: String,
}

// Update market action
#[derive(Debug, Clone, PartialEq, BorshSerialize, BorshDeserialize, Serialize, Deserialize)]
pub struct UpdateMarketAction {
    pub market: String,
    pub tick_size: Option<String>,
    pub lot_size: Option<String>,
    pub paused: bool,
    pub fee_asset: String,
}

// Orderbook depth level for aggregated view
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct OrderbookDepthLevel {
    pub price: Option<Uint128>,
    pub quantity: Option<Uint128>,
    pub order_count: u32,
}

// Manual implementation of BorshSerialize for OrderbookDepthLevel
impl BorshSerialize for OrderbookDepthLevel {
    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
        // Convert Uint128 to string for serialization
        let price_str = self.price.as_ref().map(|p| format!("{}{}", p.hi, p.lo)).unwrap_or_default();
        let quantity_str = self.quantity.as_ref().map(|q| format!("{}{}", q.hi, q.lo)).unwrap_or_default();
        
        // Serialize the string representations and order count
        BorshSerialize::serialize(&price_str, writer)?;
        BorshSerialize::serialize(&quantity_str, writer)?;
        BorshSerialize::serialize(&self.order_count, writer)?;
        
        Ok(())
    }
}

// Manual implementation of BorshDeserialize for OrderbookDepthLevel
impl BorshDeserialize for OrderbookDepthLevel {
    fn deserialize_reader<R: std::io::Read>(reader: &mut R) -> std::io::Result<Self> {
        // Deserialize the string representations and order count
        let price_str: String = BorshDeserialize::deserialize_reader(reader)?;
        let quantity_str: String = BorshDeserialize::deserialize_reader(reader)?;
        let order_count: u32 = BorshDeserialize::deserialize_reader(reader)?;
        
        // Convert strings to Uint128 if not empty
        let price = if !price_str.is_empty() {
            // Parse the string as a u128 and convert to hi/lo
            let price_val = price_str.parse::<u128>().unwrap_or_default();
            Some(Uint128 { 
                lo: price_val as u64,
                hi: (price_val >> 64) as u64,
            })
        } else {
            None
        };
        
        let quantity = if !quantity_str.is_empty() {
            // Parse the string as a u128 and convert to hi/lo
            let qty_val = quantity_str.parse::<u128>().unwrap_or_default();
            Some(Uint128 { 
                lo: qty_val as u64,
                hi: (qty_val >> 64) as u64,
            })
        } else {
            None
        };
        
        Ok(Self {
            price,
            quantity,
            order_count,
        })
    }
}

// Aggregated orderbook depth
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct OrderbookDepth {
    pub market: String,
    pub bids: Vec<OrderbookDepthLevel>,
    pub asks: Vec<OrderbookDepthLevel>,
}

// Manual implementation of BorshSerialize for OrderbookDepth
impl BorshSerialize for OrderbookDepth {
    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
        // Serialize the market string
        BorshSerialize::serialize(&self.market, writer)?;
        
        // Serialize the bids vector length
        BorshSerialize::serialize(&(self.bids.len() as u32), writer)?;
        // Serialize each bid
        for bid in &self.bids {
            BorshSerialize::serialize(bid, writer)?;
        }
        
        // Serialize the asks vector length
        BorshSerialize::serialize(&(self.asks.len() as u32), writer)?;
        // Serialize each ask
        for ask in &self.asks {
            BorshSerialize::serialize(ask, writer)?;
        }
        
        Ok(())
    }
}

// Manual implementation of BorshDeserialize for OrderbookDepth
impl BorshDeserialize for OrderbookDepth {
    fn deserialize_reader<R: std::io::Read>(reader: &mut R) -> std::io::Result<Self> {
        // Deserialize the market string
        let market: String = BorshDeserialize::deserialize_reader(reader)?;
        
        // Deserialize the bids vector
        let bids_len: u32 = BorshDeserialize::deserialize_reader(reader)?;
        let mut bids = Vec::with_capacity(bids_len as usize);
        for _ in 0..bids_len {
            bids.push(OrderbookDepthLevel::deserialize_reader(reader)?);
        }
        
        // Deserialize the asks vector
        let asks_len: u32 = BorshDeserialize::deserialize_reader(reader)?;
        let mut asks = Vec::with_capacity(asks_len as usize);
        for _ in 0..asks_len {
            asks.push(OrderbookDepthLevel::deserialize_reader(reader)?);
        }
        
        Ok(Self {
            market,
            bids,
            asks,
        })
    }
}