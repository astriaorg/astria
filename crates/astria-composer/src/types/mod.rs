pub mod rollup;
pub mod searcher;
