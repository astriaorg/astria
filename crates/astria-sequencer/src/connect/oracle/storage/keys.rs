use astria_core::connect::types::v2::{
    Base,
    CurrencyPair,
    CurrencyPairId,
    Quote,
};
use astria_eyre::eyre::{
    eyre,
    Result,
};

pub(in crate::connect::oracle) const CURRENCY_PAIR_TO_ID_PREFIX: &str =
    "connect/oracle/currency_pair_to_id/";
const ID_TO_CURRENCY_PAIR_PREFIX: &str = "connect/oracle/id_to_currency_pair/";
pub(in crate::connect::oracle) const CURRENCY_PAIR_STATE_PREFIX: &str =
    "connect/oracle/currency_pair_state/";

pub(in crate::connect::oracle) const NUM_CURRENCY_PAIRS: &str = "connect/oracle/num_currency_pairs";
pub(in crate::connect::oracle) const NUM_REMOVED_CURRENCY_PAIRS: &str =
    "connect/oracle/num_removed_currency_pairs";
pub(in crate::connect::oracle) const NEXT_CURRENCY_PAIR_ID: &str =
    "connect/oracle/next_currency_pair_id";

pub(in crate::connect::oracle) fn currency_pair_to_id(currency_pair: &CurrencyPair) -> String {
    format!(
        "{CURRENCY_PAIR_TO_ID_PREFIX}{}/{}",
        currency_pair.base(),
        currency_pair.quote()
    )
}

pub(in crate::connect::oracle) fn id_to_currency_pair(id: CurrencyPairId) -> String {
    format!("{ID_TO_CURRENCY_PAIR_PREFIX}{id}")
}

pub(in crate::connect::oracle) fn currency_pair_state(currency_pair: &CurrencyPair) -> String {
    format!("{CURRENCY_PAIR_STATE_PREFIX}{currency_pair}")
}

pub(in crate::connect::oracle) fn extract_currency_pair_from_pair_to_id_key(
    key: &str,
) -> Result<CurrencyPair> {
    extract_currency_pair_from_key(CURRENCY_PAIR_TO_ID_PREFIX, key)
}

pub(in crate::connect::oracle) fn extract_currency_pair_from_pair_state_key(
    key: &str,
) -> Result<CurrencyPair> {
    extract_currency_pair_from_key(CURRENCY_PAIR_STATE_PREFIX, key)
}

fn extract_currency_pair_from_key(prefix: &str, key: &str) -> Result<CurrencyPair> {
    let currency_pair_str = key
        .strip_prefix(prefix)
        .ok_or_else(|| eyre!("key `{key}` did not have prefix `{prefix}`"))?;
    let (base_str, quote_str) = currency_pair_str.split_once('/').ok_or_else(|| {
        eyre!("suffix `{currency_pair_str}` of key `{key}` is not of form <base>/<quote>")
    })?;
    Ok(CurrencyPair::from_parts(
        Base::unchecked_from_parts(base_str.to_string()),
        Quote::unchecked_from_parts(quote_str.to_string()),
    ))
}

#[cfg(test)]
mod tests {
    use astria_core::connect::types::v2::{
        Base,
        Quote,
    };

    use super::*;

    const COMPONENT_PREFIX: &str = "connect/oracle/";

    fn currency_pair() -> CurrencyPair {
        CurrencyPair::from_parts(
            Base::unchecked_from_parts("abc".to_string()),
            Quote::unchecked_from_parts("def".to_string()),
        )
    }

    #[test]
    fn keys_should_not_change() {
        insta::assert_snapshot!("num_currency_pairs_key", NUM_CURRENCY_PAIRS);
        insta::assert_snapshot!("num_removed_currency_pairs_key", NUM_REMOVED_CURRENCY_PAIRS);
        insta::assert_snapshot!("next_currency_pair_id_key", NEXT_CURRENCY_PAIR_ID);
        insta::assert_snapshot!(
            "currency_pair_to_id_key",
            currency_pair_to_id(&currency_pair())
        );
        insta::assert_snapshot!(
            "id_to_currency_pair_key",
            id_to_currency_pair(CurrencyPairId::new(9))
        );
        insta::assert_snapshot!(
            "currency_pair_state_key",
            currency_pair_state(&currency_pair())
        );
    }

    #[test]
    fn keys_should_have_component_prefix() {
        assert!(NUM_CURRENCY_PAIRS.starts_with(COMPONENT_PREFIX));
        assert!(NUM_REMOVED_CURRENCY_PAIRS.starts_with(COMPONENT_PREFIX));
        assert!(NEXT_CURRENCY_PAIR_ID.starts_with(COMPONENT_PREFIX));
        assert!(currency_pair_to_id(&currency_pair()).starts_with(COMPONENT_PREFIX));
        assert!(id_to_currency_pair(CurrencyPairId::new(9)).starts_with(COMPONENT_PREFIX));
        assert!(currency_pair_state(&currency_pair()).starts_with(COMPONENT_PREFIX));
    }

    #[test]
    fn should_extract_currency_pair_from_pair_to_id_key() {
        let currency_pair = currency_pair();
        let key = currency_pair_to_id(&currency_pair);
        let recovered_currency_pair = extract_currency_pair_from_pair_to_id_key(&key).unwrap();
        assert_eq!(currency_pair, recovered_currency_pair);
    }

    #[test]
    fn should_extract_currency_pair_from_pair_state_key() {
        let currency_pair = currency_pair();
        let key = currency_pair_state(&currency_pair);
        let recovered_currency_pair = extract_currency_pair_from_pair_state_key(&key).unwrap();
        assert_eq!(currency_pair, recovered_currency_pair);
    }
}
