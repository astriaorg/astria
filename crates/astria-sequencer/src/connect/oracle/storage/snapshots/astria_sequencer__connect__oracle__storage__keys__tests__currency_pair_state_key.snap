---
source: crates/astria-sequencer/src/connect/oracle/storage/keys.rs
expression: currency_pair_state(&currency_pair())
---
connect/oracle/currency_pair_state/abc/def
